app/app_main.c

```diff
@@ -3399,6 +3449,218 @@ static ACVP_RESULT app_dsa_handler(ACVP_TEST_CASE *test_case)
+            tc->d = BN_bn2hex(d);
+            tc->dlen = strlen(tc->d);
+        }
+    }
+
+    BN_CTX_free(ctx);
+    return rv;
+}
+
+static ACVP_RESULT app_kas_ecc_handler(ACVP_TEST_CASE *test_case)
+{
+    EC_GROUP *group = NULL;
+    ACVP_KAS_ECC_TC         *tc;
+    int nid = 0, exout = 0;
+    EC_KEY *ec = NULL;
+    EC_POINT *peerkey = NULL;
+    unsigned char *Z;
+    int Zlen;
+    BIGNUM *cx = NULL, *cy = NULL;
+
+    tc = test_case->tc.kas_ecc;
+
+    if (!strcmp(tc->curve, "b-233"))
+        nid = NID_sect233r1;
+    if (!strcmp(tc->curve, "b-283"))
+        nid = NID_sect283r1;
+    if (!strcmp(tc->curve, "b-409"))
+        nid = NID_sect409r1;
+    if (!strcmp(tc->curve, "b-571"))
+        nid = NID_sect571r1;
+    if (!strcmp(tc->curve, "k-233"))
+        nid = NID_sect233k1;
+    if (!strcmp(tc->curve, "k-283"))
+        nid = NID_sect283k1;
+    if (!strcmp(tc->curve, "k-409"))
+        nid = NID_sect409k1;
+    if (!strcmp(tc->curve, "k-571"))
+        nid = NID_sect571k1;
+    if (!strcmp(tc->curve, "k-224"))
+        nid = NID_secp224r1;
+    if (!strcmp(tc->curve, "p-224"))
+        nid = NID_secp224r1;
+    if (!strcmp(tc->curve, "p-256"))
+        nid = NID_X9_62_prime256v1;
+    if (!strcmp(tc->curve, "p-384"))
+        nid = NID_secp384r1;
+    if (!strcmp(tc->curve, "p-521"))
+        nid = NID_secp521r1;
+
+    if (nid == 0) {
+        printf("Invalid curve %s\n", tc->curve);
+        return ACVP_CRYPTO_MODULE_FAIL;
+    }
+
+    group = EC_GROUP_new_by_curve_name(nid);
+    if (group == NULL) {
+        printf("No group from curve name %d\n", nid);
+        return ACVP_CRYPTO_MODULE_FAIL;
+    }
+
+    ec = EC_KEY_new();
+    if (ec == NULL) {
+        printf("No EC_KEY_new\n");
+        return ACVP_CRYPTO_MODULE_FAIL;
+    }
+    EC_KEY_set_flags(ec, EC_FLAG_COFACTOR_ECDH);
+    if (!EC_KEY_set_group(ec, group)) {
+        printf("No EC_KEY_set_group\n");
+        return ACVP_CRYPTO_MODULE_FAIL;
+    }
+
+    if (!BN_hex2bn(&cx, (char *)tc->psx)) {
+        printf("BN_hex2bn failed\n");
+        return ACVP_CRYPTO_MODULE_FAIL;
+    }
+
+    if (!BN_hex2bn(&cy, (char *)tc->psy)) {
+        printf("BN_hex2bn failed\n");
+        return ACVP_CRYPTO_MODULE_FAIL;
+    }
+    peerkey = make_peer(group, cx, cy);
+    if (peerkey == NULL) {
+        printf("Peerkey failed\n");
+        return ACVP_CRYPTO_MODULE_FAIL;
+    }
+    if (!EC_KEY_generate_key(ec)) {
+        printf("EC_KEY_generate_key failed\n");
+        return ACVP_CRYPTO_MODULE_FAIL;
+    }
+
+    ec_print_key(tc, ec, 0, exout);
+    Zlen = (EC_GROUP_get_degree(group) + 7)/8;
+    if (!Zlen) {
+        printf("Zlen degree failure\n");
+        return ACVP_CRYPTO_MODULE_FAIL;
+    }
+    Z = OPENSSL_malloc(Zlen);
+    if (!Z) {
+        printf("Malloc failure\n");
+        return ACVP_CRYPTO_MODULE_FAIL;
+    }
+    if (!ECDH_compute_key(Z, Zlen, peerkey, ec, 0)) {
+        printf("ECDH_compute_key failure\n");
+        return ACVP_CRYPTO_MODULE_FAIL;
```

Comment:
Need to free all of the allocated stuff before error returns (general comment)

Reply:
got it